### PR TITLE
query HEAD name, instead of assuming main/master

### LINKED ISSUE #375: 🐛 `sl clone` silently fails for some repositories
image

Background: I am trying to clone https://github.com/cli/cli (in order to build it from source to help debug #350 )

Steps to reproduce:
1. `sl clone git@github.com:cli/cli.git`
2. `ls cli`

Expected output:
- files in https://github.com/cli/cli (`README.md`, etc.)

Actual output: 
- nothing 🙈

Other sanity-testing checks:
- `git clone` works
- `sl clone` another repo works
- `gh repo clone` works

## versions and configs
- sl: built from source on 97ec2f93b (`sl version` reports `Sapling 4.4.2`)
- git: 2.39.0
- gh: 2.21.1
- os: macOS 13.1 (Ventura)

```sh
❯ sl config
alias.commit=commit --verbose
alias.sst=!$HG status && $HG ssl
merge-tools.nvimDiff.args=-d $local $other $base -c 'redraw | echomsg "hg merge conflict, type \":cq\" to abort vimdiff"'
merge-tools.nvimDiff.executable=/usr/local/bin/nvim
merge-tools.nvimDiff.priority=100
ui.editor=nvim
ui.username=Max 👨🏽‍💻 Coplan <mchcopl@gmail.com>
```

```sh
git config --list

credential.helper=osxkeychain
alias.s=!{ git for-each-ref --sort=-committerdate --format='%(refname)' refs/heads/ refs/remotes/$(git remote || printf origin)|sed -e 's-refs/heads/--'| sed -e "s-refs/remotes/$(git remote || printf origin)/--" & printf 'develop
main
master
'; } | uniq | grep -i -m1 --color=never $1 | xargs -I {} sh -c " echo {}; git switch {}"
alias.sf=!git switch $(git for-each-ref --sort=-committerdate --format='%(refname)' refs/heads/ refs/remotes/$(git remote || printf origin)/|sed -e 's-refs/heads/--'| sed -e "s-refs/remotes/$(git remote || printf origin)/--" | uniq | fzf) $@
alias.fr=!git fetch $(git remote | head) && git rebase
alias.frm=!git fetch $(git remote | head) && git rebase $(git remote | head)/main
alias.fru=!git fetch upstream && git rebase upstream/main
alias.cw=commit --fixup=wip
core.pager=${_GIT_CONTRIB:-/usr/local/share/git-core/contrib}/diff-highlight/diff-highlight | $PAGER
color.branch.current=yellow reverse
color.branch.local=yellow
color.branch.remote=green
color.diff.meta=yellow bold
color.diff.frag=magenta bold
color.status.added=green bold
color.status.changed=red bold strike
color.status.untracked=cyan
color.status.branch=yellow bold ul
commit.verbose=true
interactive.difffilter=$_GIT_CONTRIB/diff-highlight/diff-highlight | $PAGER
rebase.instructionformat=(%al) %s
pull.fastforward=true
pull.ff=only
pull.rebase=true
user.email=Max.Coplan@walmart.com
user.name=Max 👨🏽‍💻 Coplan
help.autocorrect=1
url.git@github.com:.insteadof=https://github.com/
```

```sh
❯ gh config list
git_protocol=ssh
editor=
prompt=enabled
pager=
http_unix_socket=
browser=
```

Please let me know what other info you need 🙂 

Happy New Year 🎆 🎊 🎈

### PR DESCRIPTION
**Pull request by discentem**
query HEAD name, instead of assuming main/master

Summary:

Test Plan:

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/facebook/sapling/pull/386).
* __->__ #386

### CODE REVIEW

File: eden/scm/edenscm/git.py [hunk 1]
```diff
@@ -142,7 +143,24 @@ def clone(ui, url, destpath=None, update=True, pullnames=None):
 
                 if not nodes:
                     pullnames.append(update)
-
+            else:
+                b = callgitnorepo(ui, ["ls-remote", "--symref", url, "HEAD"])
+                def parse_symref_head(b):
```
Comment (discentem) [inline eden/scm/edenscm/git.py:148]: I'm sure there is a better way to do this. Happy to make this cleaner, move the logic to a class, etc based on suggestions. 

At first I tried to use the [RefName](https://github.com/facebook/sapling/blob/5f017b9a2080de77130458d7168d23a886586432/eden/scm/edenscm/git.py#L273) class but it doesn't seem like there is any existing method that would yield the default branch name from the full ref.

Comment (vegerot) [reply eden/scm/edenscm/git.py:148]: a(n unhelpful) review of your PR is at https://www.twitch.tv/videos/1697938402?t=03h52m12s

Comment (discentem) [reply eden/scm/edenscm/git.py:148]: I don't know whether to feel embarrassed or proud that someone live reviewed my PR on Twitch 😂 😳 😅

Comment (discentem) [reply eden/scm/edenscm/git.py:148]: (@vegerot BTW I am curious but have no context as to why someone asked it to be reviewed this stream? It wasn't me haha as I was not watching. Anyway, thanks! Or was it [strager](https://github.com/strager)? And yes I only meant that I thought there might be a better way to parse text 😆)

Comment (vegerot) [reply eden/scm/edenscm/git.py:148]: umm, since the stream is http://twitch.tv/strager, I'd assume @strager was the reviewer

**Review by vegerot — COMMENTED** [review]
It would be nice if you added some automated tests for the symref parsing

File: eden/scm/edenscm/git.py [hunk 2]
```diff
@@ -142,7 +143,24 @@ def clone(ui, url, destpath=None, update=True, pullnames=None):
...
+                    82156f7fd2719f5cca9f8a8738ff4e61561d04aa        HEAD
+                    """
+                    symrefhead = out.split("\n")[0]
+                    '''
+                    ref: refs/heads/trunk   HEAD
+                    '''
+                    headref = symrefhead.split(" ")[1].split("\t")[0]
+                    '''
+                    refs/heads/trunk
+                    '''
```
Comment (vegerot) [inline eden/scm/edenscm/git.py:161]: nit: your comment style are inconsistent.  You sometimes use `"""` and sometimes use `'''`

Comment (discentem) [reply eden/scm/edenscm/git.py:161]: Good call out, will fix this

**Comment by discentem** [comment]
> It would be nice if you added some automated tests for the symref parsing

Happy to! Can you suggest where to add these tests? I'm not yet familiar with the test structure or organization in this project.

**Comment by vegerot** [comment]
> > It would be nice if you added some automated tests for the symref parsing
> 
> Happy to! Can you suggest where to add these tests? I'm not yet familiar with the test structure or organization in this project.

Just kidding!  The automated tests don’t work for the OSS build.  I’ve complained about this before, but for this PR I think the simplest thing to do is get a job at Meta and use their tools to run the test suite

File: eden/scm/edenscm/git.py (hunk 2 above)
Comment (discentem) [reply eden/scm/edenscm/git.py]: fixed

**Comment by vegerot** [comment]
> > > > It would be nice if you added some automated tests for the symref parsing
> > > 
> > > 
> > > Happy to! Can you suggest where to add these tests? I'm not yet familiar with the test structure or organization in this project.
> > 
> > 
> > Just kidding! The automated tests don’t work for the OSS build. I’ve complained about this before, but for this PR I think the simplest thing to do is get a job at Meta and use their tools to run the test suite
> 
> 😂
> 
> I should probably add a unit test for this somewhere though at least 🤔

wouldn't bother until #447 is addressed

File: eden/scm/edenscm/git.py [hunk 3]
```diff
@@ -142,7 +143,25 @@ def clone(ui, url, destpath=None, update=True, pullnames=None):
 
                 if not nodes:
                     pullnames.append(update)
-
+            else:
+                b = callgitnorepo(ui, ["ls-remote", "--symref", url, "HEAD"])
+                b.check_returncode()
+                def parse_symref_head(b) -> str:
+                    out = b.stdout.decode('utf-8')
+                    """
```
Comment (quark-zju) [inline eden/scm/edenscm/git.py:151]: nit: we don't use docstrings inside function body like this. Consider using `#` comments and/or making it a doctest. Something like:

    def parse_symref_head(git_output: str) -> Optional[str]:
        """Extract the branch name referred by HEAD.
        Return None if the branch name cannot be detected.
        git_output is the output of 'git ls-remote --symref url HEAD'.

        >>> parse_symref_head('ref: refs/heads/main\tHEAD\nd0a9bf5e93f6f6828d4b7e3ff740c099d2ca6866\tHEAD\n')
        'main'
        >>> parse_symref_head('')
        None
        >>> parse_symref_head('ref: refs/tags/main\tHEAD\nd0a9bf5e93f6f6828d4b7e3ff740c099d2ca6866\tHEAD\n')
        None
        """

Adding `testmod("edenscm.ext.remotenames")` to `tests/test-doctest.py` would include the doctest in tests.

The function should handle cases where HEAD does not exist or is not a branch (ex. prefix is not 'refs/heads'). Some tests actually rely on non-existed HEAD so they currently break with this change.

Comment (discentem) [reply eden/scm/edenscm/git.py:151]: Hey @quark-zju, thanks for the reply! Can you share more info about doctests? I'm not familiar with how they work. 

Also what should Sapling do if parse_symref_head returns none -- is it expected to abort?

Comment (bolinfest) [reply eden/scm/edenscm/git.py:151]: @discentem I'll work on cleaning this up now that I have it imported on our side.

Comment (discentem) [reply eden/scm/edenscm/git.py:151]: Great! Thanks @bolinfest 😀

**Comment by bolinfest** [comment]
In the end, https://github.com/facebook/sapling/commit/c8b66c68991066b3e8731630ef757e94cd03ff36 kept the general idea, but you can see how I added doctests and changed the ordering of where things happened within `clone()`. It was a good sign that there was only one small change to a `.t` test!

**Pull request closed without merging**